service/history/timerBuilder.go

```diff
@@ -174,35 +174,33 @@ func (tb *timerBuilder) AddActivityTimeoutTask(scheduleID int64,
 }
 
 // AddUserTimer - Adds an user timeout request.
-func (tb *timerBuilder) AddUserTimer(ti *persistence.TimerInfo, msBuilder *mutableStateBuilder) persistence.Task {
+func (tb *timerBuilder) AddUserTimer(ti *persistence.TimerInfo) {
 	tb.logger.Debugf("Adding User Timeout for timer ID: %s", ti.TimerID)
+	tb.loadUserTimer(ti.ExpiryTime, ti.TimerID, ti.TaskID != emptyTimerID)
+}
 
-	// TODO: This is broken.  We need to comeup with a better way to implement this
-	tb.LoadUserTimers(msBuilder)
+// LoadUserTimers - Load all user timers from mutable state.
+func (tb *timerBuilder) LoadUserTimers(msBuilder *mutableStateBuilder) {
+	tb.userTimers = timers{}
+	tb.pendingUserTimers = msBuilder.pendingTimerInfoIDs
+	for _, v := range msBuilder.pendingTimerInfoIDs {
```

Comment:
optimization: when you are loading the timers from mutable state for the first time, it would be much more efficient to just allocate a slice of len(pendingTimerInfoIDs) and maybe, even call Sort() once.